### ISSUE TITLE
Keycloak does not sign SAML logout responses

### ISSUE THREAD

**Issue by andreas-kupries** [keycloak/keycloak#30680]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

saml

### Describe the bug

Using KC 25.0.1, the latest version of KC at the time of writing this.

Attempting a SAML Single LogOut flow the flow is aborted in the SP on receiving KC's logout response.

The `crewjam/saml` package reports `signature element not present`. Capturing the response with the `saml tracer` browser plugin confirms that KC has not signed its SAML logout response.

As far as I know the crewjam error report is correct, as SAML logout responses MUST be signed.
Ref https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf section 4.4.3.4, page 36, lines 1276ff
```
The <LogoutResponse> message MUST be signed if the HTTP POST or Redirect binding is used
```
My setup uses post/redirect bindings.

In the discussion https://github.com/keycloak/keycloak/discussions/30618 I used 23.0.4, and asked if it missed a configuration setting.

Checked the 25.0.1 for this and AFAICS no changes in the UI, nor in the behaviour, and also no responses to my discussion query.

Thus escalating this now to bug status

### Version

25.0.1

### Regression

- [ ] The issue is a regression

### Expected behavior

For KC to sign the SAML logout responses.

### Actual behavior

KC does not sign the SAML logout responses.

### How to Reproduce?

As far as I can tell, just create a KC realm, configure with a SAML client. 
Attached is a json export of that client's configuration.

[local-rancher(4).json](https://github.com/user-attachments/files/15951644/local-rancher.4.json)

If there is something missing, well, nobody answered my question in discussion
 https://github.com/keycloak/keycloak/discussions/30618

**Comment by rmartinc** [comment]
I cannot reproduce this in my env. How are you initiating the single logout in keycloak? Maybe I'm doing something different. In my case the logout from the keycloak is always signed no matter if it's POST or REDIRECT.

Is it possible to attach a har from the browser of the logout process?

**Comment by andreas-kupries** [comment]
Thank you for checking. While I am off work in my TZ right now I will look into making a HAR tomorrow ASAP.
Saw the notification despite that because my personal email is the default for any unknown repos ;)

**Comment by andreas-kupries** [comment]
Something I failed to provide, with apologies for my forgetfulness here, the local keycloak I test against is launched via

```
#!/bin/bash
docker run -p 9999:8080 \
       -e KEYCLOAK_ADMIN=admin \
       -e KEYCLOAK_ADMIN_PASSWORD=admin \
       -e KEYCLOAK_LOGLEVEL=DEBUG \
       -e KC_LOG_LEVEL=DEBUG \
       quay.io/keycloak/keycloak:25.0.1 \
       start-dev \
    |& tee kc.log
```

**Comment by andreas-kupries** [comment]
[tagetarl_Archive [24-06-26 09-35-51].zip](https://github.com/user-attachments/files/15984212/tagetarl_Archive.24-06-26.09-35-51.zip)

And here is a HAR file of the logout process in a zip.
Hopefully that helps.

Changed the email routing as well, hopefully get further notifications at the work mail.
I.e. enabling me to respond faster.

**Comment by rmartinc** [comment]
Thanks @andreas-kupries!

If you see the response from keycloak is signed. It's a redirect and the location in the response is:

```
https://tagetarl:8005/v1-saml/keycloak/saml/slo?SAMLResponse=jVLBasMwDP2V4ntqN2mS2iSBsV4K3WUtPewyFEdpQx27WM7Y%2Fn5pusEGo0y3h96TeE8qCHpzUVt3dEN4Rro4Szh7740lNbVKNnirHFBHykKPpIJWu4enrYrnQl28C047w26S%2B2QgQh86Z9mPBf%2BWrJFCZ%2BEKSnYK4UKK8wBHDOCNWgmR8rdFdB3Jz%2FihjYMznxAZx2abdck261dRg2jTXEd1LWW0lHkcyaWuI4G6hSxpEr2QI9l%2BJ7F3JeuaKGnzWjcylU2OrWixzfLVIknqVYNZhosc5UqCwHSUEg24sRTAhpLFIl5GIovibC9ylaQqyecyWb6w2QE9TU5Gl6wqJpmvrrZGV8ZpMCdHQcmxuEcwPXEPVp%2FQF%2FyLXNxOtwsQBvqNHl2DswOYAe%2FHSxNb7QatkYjxquC%2Fh%2FK%2F3qP6BA%3D%3D&RelayState=HN5rxfCr1is99R3gpndF8w-MQL7tomvC_a1T50iDNLmAGE3kXtEXaChX&SigAlg=http%3A%2F%2Fwww.w3.org%2F2001%2F04%2Fxmldsig-more%23rsa-sha256&Signature=Cjub%2FMYvAEWIhG5VKmW93r8CC3DCPQKmd2oUdgcbJgXQ1M4CBGjvHIYsbehPEPkQRJJJ9E%2BJDZHd3y%2BD4UZptb6AVP6kMQ78nh5UfgVm7p1G6MtTvVTZB2RJrkynlAYrzX89PXQS7NKgMZGootXTKHBQrPiH084CFpIeiJqeejjX7qWA2SzmGyfeNeOp1utW%2FG0qnsUa2H%2BilSRMimfbGBHvZJfTA0UiHB17i%2BDLex3IbHCUzw78iNgjrgSvO5NKYYBJxtZNY6%2BDtQq6GD8KDXZ%2FpXopWTIx6myk2Un0Sn9lK49nP823A2E1tftDm1gJm6FnaHMdXlylFZm6lIbr7w%3D%3D
```

If you there is a `SAMLResponse`, a `RelayState`, a `SigAlg` and `Signature` query parameters in the URL. So the redirect is signed and everything is OK.

My feeling is that the client (rancher) is not managing the signatures in the redirect binding OK. I think they are adding the signature inside the document as it's done for POST. If you check the [SAML spec](https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf) (chapter 3.4.4.1 DEFLATE Encoding), XML signatures are performed differently in REDIRECT. The signature  is not added in the document response but in an additional query string parameter. Checking the rancher request to keycloak they are signing inside the document and not using other query parameters.

In summary, the client is not singing OK in REDIRECT binding. You can try to configure POST (if possible) or talk with the client maintainers and ask about this.

**Comment by andreas-kupries** [comment]
Thank you for the explanations.
The underlying client is https://github.com/crewjam/saml
And yes, it searches for the signature in the XML document itself. And the error report is about not finding it there.

The package can apparently send requests with `SigAlg`, but only for Auth redirect, not Logout, and reception does not look for SigAlg etc. at all :frowning: 

We may have to patch crewjam. (Maintenance seems to have ceased).
Or find something better / more official to handle SAML in go.
(Do you know such a thing ?)

**Comment by andreas-kupries** [comment]
Thanks again for putting me on the right track here. Further searching in the crewjam issues and prs I may have found a possible fix waiting for merge - https://github.com/crewjam/saml/pull/449

Feel free to close this ticket, and I will now work on patching our local copy of this package (we already had to fork it to apply a patch fixing mishandling of compressed data).

**Comment by rmartinc** [comment]
No, sorry, no idea. But if the client support POST configuration maybe you can override the issue (besides POST is always more secure as the data in in the payload).

I'm closing this as the issue seems to be at client side.

**Comment by andreas-kupries** [comment]
Application of https://github.com/crewjam/saml/pull/449 + https://github.com/crewjam/saml/pull/449/files#r1654477177
gave me success.
